.gitlab-ci-darwin.yml

```diff
@@ -1,26 +1,49 @@
 variables:
-  SCHEDULER_PARAMETERS: '--nodes=1 --partition=power9 --export=NONE'
+  SCHEDULER_PARAMETERS: '--nodes=1 --partition=power9-eap-ci --export=NONE'
   GIT_SUBMODULE_STRATEGY: recursive
 
 stages:
+  - performance-target
   - performance-regression
 
 .gcc-mpi-cuda-performance-regression:
   variables: 
     CMAKE_CXX_COMPILER: $CI_PROJECT_DIR/external/Kokkos/bin/nvcc_wrapper
   script:
-    - env -i bash --norc --noprofile ./scripts/darwin/build_fast.sh
+    - env
+    - git fetch --all
+    - printf 'env -i bash --norc --noprofile %s/scripts/darwin/build_fast_target.sh  "%s" "%s" "%s" "%s"\n' ${CI_PROJECT_DIR} ${GITHUB_APP_PEM} ${CI_COMMIT_SHA} ${CI_COMMIT_BRANCH} ${CI_JOB_URL}> execute_main.sh
+    - printf 'env -i bash --norc --noprofile %s/scripts/darwin/metrics.sh "%s" "%s/build" "%s" "%s"\n' ${CI_PROJECT_DIR} ${GITHUB_APP_PEM} ${CI_PROJECT_DIR} ${CI_COMMIT_SHA} ${CI_COMMIT_BRANCH}>> execute_main.sh
+    - |
+      STATUS=$(env -i bash --norc --noprofile ${CI_PROJECT_DIR}/scripts/darwin/build_fast.sh "${GITHUB_APP_PEM}" "${CI_COMMIT_SHA}" "${CI_COMMIT_BRANCH}" "${CI_JOB_URL}")
+      echo "PASS_OR_FAIL ${STATUS}"
+      if [[ "$STATUS" == *"FAILED"* ]]; then exit 1; fi
+    - |
+      STATUS=$(env -i bash --norc --noprofile ${CI_PROJECT_DIR}/scripts/darwin/metrics.sh "${GITHUB_APP_PEM}" "${CI_PROJECT_DIR}/build" "${CI_COMMIT_SHA}" "${CI_COMMIT_BRANCH}")
+      echo "PASS_OR_FAIL ${STATUS}"
+      if [[ "$STATUS" == *"FAILED"* ]]; then exit 1; fi
 
-  artifacts:
-    expire_in: 3 days
-    paths:
-      - ${CI_PROJECT_DIR}/build/tst/regression/outputs/advection_performance/performance.png
-      - ${CI_PROJECT_DIR}/build/tst/regression/outputs/advection_performance_mpi/performance.png
+.gcc-mpi-cuda-performance-regression-target-branch:
+  script:
+    - git fetch --all
+    - printf 'env -i bash --norc --noprofile %s/scripts/darwin/build_fast_target.sh  "%s" "%s/build" "%s" "%s" "%s %s"\n' ${CI_PROJECT_DIR} ${GITHUB_APP_PEM} ${CI_PROJECT_DIR} ${CI_COMMIT_SHA} ${CI_COMMIT_BRANCH} ${CI_JOB_URL} ${CI_JOB_TOKEN} > execute_target.sh
+    - |
+      STATUS=$(env -i bash --norc --noprofile ${CI_PROJECT_DIR}/scripts/darwin/build_fast_target.sh "${GITHUB_APP_PEM}" "${CI_PROJECT_DIR}/build"  "${CI_COMMIT_SHA}" "${CI_COMMIT_BRANCH}" "${CI_JOB_URL}" "${CI_JOB_TOKEN}")
+      echo "PASS_OR_FAIL ${STATUS}"
+      if [[ "$STATUS" == *"FAILED"* ]]; then exit 1; fi
```

Comment:
So it turns out this is a bit difficult to make DRY. The ci yaml files do not really have a notion of functions. It would be nice if I could just put the logic in the shell script for handling the error codes but this does not work because when I call the script using the "env" command the error code gets lost. It is important that the right exit code is called from within the yaml file because this is how we know if there was a problem or not. At this point as nasty as it seems I will be leaving the "if logic" as is.

Reply:
Ok, I figured out how to do this with anchors and templates.